### PR TITLE
DBZ-191 support default value of column in mysql

### PR DESCRIPTION
**Pull request by sweat123**
@gunnarmorling @jpechane 

a minimum implementation; 
Hope to giving some guidance and suggestions;

### CODE REVIEW

**Comment by sweat123** [comment]
This PR is a minimum implementation;  Just want to verify that the way parsing default value is correct or not and hope any one can give me some guidance and suggestions; 
Thanks!

File: debezium-core/src/main/java/io/debezium/relational/ColumnEditorImpl.java [hunk 1]
```diff
@@ -22,6 +22,8 @@
     private boolean optional = true;
     private boolean autoIncremented = false;
     private boolean generated = false;
+    private Object defaultValue = null;
+    private boolean isDefaultValueNull = false;
```
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/relational/ColumnEditorImpl.java]: Why do you need this attribute? Cannot be it be computed on-the-fly just by comparing `defaultValue == null` in `isDefaultValueNull()` method?

File: debezium-core/src/main/java/io/debezium/time/ZonedTimestamp.java [hunk 2]
```diff
@@ -90,10 +94,26 @@ public static String toIsoString(Object value, ZoneId defaultZone, TemporalAdjus
...
+        if (value instanceof String && isTimestampFormat((String) value)) {
+            return (String) value;
+        }
         throw new IllegalArgumentException(
                 "Unable to convert to OffsetDateTime from unexpected value '" + value + "' of type " + value.getClass().getName());
     }
 
+    public static boolean isTimestampFormat(String timestamp) {
+        try {
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
```
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/time/ZonedTimestamp.java]: Is the timestamp format always the same or is dependent on database settings?

File: debezium-core/src/main/java/io/debezium/relational/ColumnEditorImpl.java (hunk 1 above)
Comment (sweat123) [reply debezium-core/src/main/java/io/debezium/relational/ColumnEditorImpl.java]: If the default value in ddl is not set, the `defaultValue` is `null` and we should not set `null` at field which in schema `defaultValue`;
If the default value in ddl is set with `null`, we need to convert `null` to schema type and set schema `defaultValue` with value which is converted;

>ALTER TABLE aaatest123 ADD COLUMN ddd INT DEFAULT NULL;

The default value of column `ddd` is null in ddl, we will set `isDefaultValueNull` true and then convert null to schema type;

>ALTER TABLE aaatest123 ADD COLUMN ddd INT;

The default value of column `ddd` is not defined in ddl, we will skip this column and the field `defaultValue` in `ConnectSchema` will not be set;

**Comment by MrTrustworthy** [comment]
I tested the current master with this branch merged in on a MySQL 5.6 with various defaults and got the following error:

```
Caused by: io.debezium.text.ParsingException: Unexpected exception while parsing statement CREATE TABLE "my_table" (
	...
	"my_date" datetime NOT NULL DEFAULT '0000-00-00 00:00:00',
	...
Caused by: java.lang.IllegalArgumentException: Unexpected value for JDBC type 93 and column my_date DATETIME NOT NULL DEFAULT VALUE 0000-00-00 00:00:00: class=class java.lang.String
	at io.debezium.jdbc.JdbcValueConverters.handleUnknownData(JdbcValueConverters.java:1242)
	at io.debezium.jdbc.JdbcValueConverters.convertTimestampToEpochMillis(JdbcValueConverters.java:424)
	at io.debezium.jdbc.JdbcValueConverters.lambda$converter$20(JdbcValueConverters.java:302)
	at io.debezium.connector.mysql.MySqlDdlParser.convertDefaultValueToSchemaType(MySqlDdlParser.java:797)
	at io.debezium.connector.mysql.MySqlDdlParser.parseCreateColumn(MySqlDdlParser.java:759)
	at io.debezium.connector.mysql.MySqlDdlParser.parseCreateDefinition(MySqlDdlParser.java:735)
	at io.debezium.connector.mysql.MySqlDdlParser.parseCreateDefinitionList(MySqlDdlParser.java:583)
	at io.debezium.connector.mysql.MySqlDdlParser.parseCreateTable(MySqlDdlParser.java:395)
	at io.debezium.connector.mysql.MySqlDdlParser.parseCreate(MySqlDdlParser.java:298)
	at io.debezium.connector.mysql.MySqlDdlParser.parseNextStatement(MySqlDdlParser.java:169)
	at io.debezium.relational.ddl.DdlParser.parse(DdlParser.java:286)
```

Not exactly sure why this happened. The default value for date is actually exactly in the format that @jpechane asked about in the review above.

**Comment by MrTrustworthy** [comment]
To add onto my previous comment:

Seems like Conversions.java:toLocalDateTime() does not expect a String but instead wants the input to be a date type. Conversely, MySqlDdlParser.java:convertDefaultValueToSchemaType() passes the string "0000-00-00 00:00:00" down to the various converters even though the JDBCType is Types.TIMESTAMP. 
Should probably convert that string to a valid java time object at some point along the way, but i'm not exactly sure what the right place would be.

**Comment by sweat123** [comment]
@MrTrustworthy You are right; 

string "0000-00-00 00:00:00" will not appear in binlog, so the input data in `toLocalDateTime()` will not get string type and can work correctly; When I have add `MySqlValueConverters.java` in `MySqlDdlParser`, `MySqlValueConverters` have to handle much more input data type; 

It seems I have to add more method to convert multiple input data type ;

**Comment by MrTrustworthy** [comment]
The issue seems to be a bit more complex, and I'm not entirely sure that I completely understand the approach that debezium uses to serialize default values. Using your last commits, there are warnings in the MySqlValueConverters.java:

`WARN  MySqlValueConverters:659 - Unexpected JDBC DATE value for field my_date with schema org.apache.kafka.connect.data.SchemaBuilder@5286a6e0: class=class java.lang.String, value=0000-00-00`

This causes an error later in TableSchemaBuilder.addField():309 because the schema builder expects an INT64 instead of a String as default value.

**Comment by sweat123** [comment]
@MrTrustworthy 
em.. I have not a good idea how to convert date type correctly; And in my submited code, I have not support converting date, timestamp and other date type;

**Comment by jpechane** [comment]
@sweat123 It seems it is necessary to add a support for text to type conversion to all values supported by MySQL. For formatting dependent types I propose to delegate the conversion to a separate utility class. In the first step it would just parse the default formatting, in the next one we should enable user to set the format. It would be similar to what we have right now for wal2json plugin in PostgreSQL.

**Comment by jpechane** [comment]
@sweat123 I've pushed changes into the branch
* hadnling of default nulls is simlified
* a conversion for some of datatypes was added

Please take a look if it is ok for you. If you are still interested in working on PR it would be great if you could verify all supported datatypes and add converter for those that requires it.

**Comment by jpechane** [comment]
For the record - we expect this PR to be put on hold for a while - https://gitter.im/debezium/dev?at=5ae2c16862316e0505055d73

**Comment by sweat123** [comment]
https://github.com/debezium/debezium/pull/510  new branch was created. Current branch  can't now pass the integration test due to `UnwrapFromMongoDbEnvelopeTestIT.class`

**Pull request closed without merging**